memory.x

```diff
@@ -13,21 +13,31 @@ SECTIONS {
     .boot2 ORIGIN(BOOT2) : {
         KEEP(*(.boot2));
     } > BOOT2
+} INSERT BEFORE .text;
 
+SECTIONS {
     /* ### Main ram section */
-    /* .ram ORIGIN(RAM) : { TODO: FIXME: IDK HOW LINKERS WORK
-        *(.ram.*)
+    .ram : {
+        *(.ram .ram.*)
         . = ALIGN(4);
-    } > RAM */
+    } > RAM AT > FLASH
+} INSERT AFTER .data;
 
+SECTIONS {
     /* ### Small 4kb memory sections for high bandwidth code or data per core */
-    /* .small.0 ORIGIN(SMALL0) : { FIXME: THESE CAUSE EVERYTHING BUT PICO-DVI TO RUN
-        *(.small.0.*)
+    .scratch_x : {
+        _scratch_x_start = .;
+        *(.scratch_x .scratch_x.*)
         . = ALIGN(4);
-    } > SMALL0
+        _scratch_x_end = .;
+    } > SCRATCH_X AT > FLASH
+    _scratch_x_source = LOADADDR(.scratch_x);
```

Comment:
Just so I am understanding this right, `_scratch_x_start` will point to the memory address in SCRATCH_X of the .scratch_x region, and the `_scratch_x_source` will point to the start of the memory location where the data is stored in flash

Reply:
Yes. These are symbols provided by the linker and made available to code, which in turn will use them to copy the initial data from flash to RAM. The code is compiled assuming the RAM addresses, for example if you say `.4byte function_pointer` that will be the address in RAM rather than flash. [This thread](https://www.embeddedrelated.com/showthread/comp.arch.embedded/77071-1.php) explains that in a bit of detail, though instead of "crt0" read "cortex-m-rt".